src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -567,6 +567,24 @@ public T[] ToArray()
         }
 
         /// <summary>
+        /// Removes all values from the <see cref="ConcurrentBag{T}"/>.
+        /// </summary>
+        public void Clear()
+        {
+            // Short path if the bag is empty
+            if (_headList == null)
+                return;
+
+            // Acquire the global lock to update the pointers
+            lock (GlobalListsLock)
+            {
+                _locals = new ThreadLocal<ThreadLocalList>();
+                _headList = null;
+                _tailList = null;
```

Comment:
You don't right.
Method FreezeBag used by read only methods such like: CopyTo, ToArray, GetEnumerator, Count and IsEmpty.
Methods that change bag use `lock (GlobalListsLock)` see `GetThreadList` method that used by next methods: Initialize, Add, TryTakeOrPeek.
So my lock is correct. 
ThreadLocal class do not have methods to clear. So we must recreate it. After that all local lists will be garbage by GC because we do have any reference to them.